go/store/types/json.go

```
@@ -0,0 +1,387 @@
+
+	return JSON{
+		valueImpl{
+			buff:    buff,
+			offsets: offsets,
+			vrw:     vrw,
+			nbf:     t.nbf,
+		},
+	}
+}
+
+func (t JSON) Empty() bool {
+	return t.Len() == 0
+}
+
+func (t JSON) Format() *NomsBinFormat {
+	return t.format()
+}
+
+// Value interface
+func (t JSON) Value(ctx context.Context) (Value, error) {
+	return t, nil
+}
+
+func (t JSON) Inner() (Value, error) {
+	dec := newValueDecoder(t.buff, t.vrw)
+	dec.skipKind()
+	dec.skipCount()
+	return dec.readValue(t.nbf)
+}
+
+func (t JSON) WalkValues(ctx context.Context, cb ValueCallback) error {
+	val, err := t.Inner()
+	if err != nil {
+		return err
+	}
+	return val.WalkValues(ctx, cb)
+}
+
+func (t JSON) typeOf() (*Type, error) {
+	val, err := t.Inner()
+	if err != nil {
+		return nil, err
+	}
+	return val.typeOf()
+}
+
+func (t JSON) Kind() NomsKind {
+	return JSONKind
+}
+
+func (t JSON) decoderSkipToFields() (valueDecoder, uint64) {
+	dec := t.decoder()
+	dec.skipKind()
+	count := dec.readCount()
+	return dec, count
+}
+
+func (t JSON) Len() uint64 {
+	return 1
+}
+
+func (t JSON) isPrimitive() bool {
+	return false
+}
+
+func (t JSON) Less(nbf *NomsBinFormat, other LesserValuable) (bool, error) {
+	otherJSONDoc, ok := other.(JSON)
+	if !ok {
+		return JSONKind < other.Kind(), nil
+	}
+
+	cmp, err := t.Compare(otherJSONDoc)
+	if err != nil {
+		return false, err
+	}
+
+	return cmp == -1, nil
+}
+
+// Compare implements MySQL JSON type compare semantics.
+func (t JSON) Compare(other JSON) (int, error) {
+	left, err := t.Inner()
+	if err != nil {
+		return 0, err
+	}
+
+	right, err := other.Inner()
+	if err != nil {
+		return 0, err
+	}
+
+	return compareJSON(left, right)
+}
+
+func (t JSON) readFrom(nbf *NomsBinFormat, b *binaryNomsReader) (Value, error) {
+	panic("unreachable")
+}
+
+func (t JSON) skip(nbf *NomsBinFormat, b *binaryNomsReader) {
+	panic("unreachable")
+}
+
+func (t JSON) HumanReadableString() string {
+	val, err := t.Inner()
+	if err != nil {
+		d.PanicIfError(err)
```

I can imagine a bug where the server is having issues and we log the human readable string, only to find out that the json was corrupt and instead of logging we panic.